### PR TITLE
Read config file function for the "main" files

### PR DESCRIPTION
**Pull request by memem359**
My third attempt (after #1519 and #1520 ) to do (what was in theory) a simple refactoring of the camain, chmain, and dmain files. There was a To-Do comment in those files:

> TODO Code combining config, persistent_config and commandline args is copy-pasted slightly differently in chmain, dmain and camain. Make it into a single function.

I tried a single player game, and didn't notice any changes (as expected).

@LGM-Doyle made a suggestion that was better/simpler than what I was trying, so hopefully the third PR will be the last...

### CODE REVIEW

File: util/OptionsDB.cpp
Comment (geoffthemedio) [inline util/OptionsDB.cpp]: de-indent { and } to match if, and only single-tab indent the enclosed code

File: util/OptionsDB.cpp
Comment (geoffthemedio) [inline util/OptionsDB.cpp]: signature probably long enough to break over lines

File: util/OptionsDB.cpp
Comment (geoffthemedio) [inline util/OptionsDB.cpp]: indent << to match that of previous line

File: util/OptionsDB.h
Comment (geoffthemedio) [inline util/OptionsDB.h]: weird indenting before of

File: server/dmain.cpp
Comment (adrianbroher) [inline server/dmain.cpp]: Whitespace.

File: server/dmain.cpp
Comment (adrianbroher) [inline server/dmain.cpp]: Whitespace.

File: client/AI/camain.cpp
Comment (adrianbroher) [inline client/AI/camain.cpp]: Whitespace.

File: util/OptionsDB.cpp
Comment (adrianbroher) [inline util/OptionsDB.cpp]: Use `file_path` instead of `GetPersistentConfigPath`.

File: util/OptionsDB.cpp
Comment (adrianbroher) [inline util/OptionsDB.cpp]: Bracket not on same line as definition.

File: util/OptionsDB.cpp
Comment (adrianbroher) [inline util/OptionsDB.cpp]: Irrelevant comment here but missing in the function documentation.

File: util/OptionsDB.cpp
Comment (adrianbroher) [inline util/OptionsDB.cpp]: Ew, fix that indentation.

File: util/OptionsDB.h
Comment (adrianbroher) [inline util/OptionsDB.h]: No description of what this function does, nor the parameters nor the point of the default value of version.

File: util/OptionsDB.h
Comment (adrianbroher) [inline util/OptionsDB.h]: Pointless space between return type and function name.

**Review by adrianbroher — CHANGES_REQUESTED** [review]
Other issues:

  * Author doesn't match committer.
  * When fixing issues please stop creating PR's, push to existing PR's instead.  Force push to clean the PR up with rebasing after all problems with the PR are handled.

File: util/OptionsDB.h
Comment (geoffthemedio) [reply util/OptionsDB.h]: it's consistent with the surrounding code

File: util/OptionsDB.cpp
Comment (memem359) [reply util/OptionsDB.cpp]: This was copied over from the previous code (chmain.cpp), but easily taken out.

File: util/OptionsDB.cpp
Comment (memem359) [reply util/OptionsDB.cpp]: Missed that one, thanks.

File: util/OptionsDB.h
Comment (memem359) [reply util/OptionsDB.h]: That was exactly it. (I copied the next lines from SetFromXML, then reworded it.)

**Comment by memem359** [comment]
That was painful (for a git novice), but I think I managed to get all of the requested changes by geoff and adrian completed (except for the "pointless space between return type and function name", which was done to match the rest of the header style).

File: util/OptionsDB.cpp
Comment (geoffthemedio) [inline util/OptionsDB.cpp]: { to next line

File: util/OptionsDB.cpp
Comment (geoffthemedio) [inline util/OptionsDB.cpp]: { to next line

**Review by geoffthemedio — DISMISSED** [review]

File: util/OptionsDB.cpp
Comment (memem359) [reply util/OptionsDB.cpp]: The FO Code Standards wiki says to follow your preference, as long as it doesn't contradict the style of the rest of the file.

Might be useful to mention a preferred (or prevalent) style in the Wiki, when it isn't obvious from a given file.

Comment (geoffthemedio) [reply util/OptionsDB.cpp]: For multi-line function definitions, the { should always be on a line of its own.

You can put the { at the end of the definition line if it fits in one line, but not if it's split over multiple lines.

Applies for if, loop, and function definition blocks.

File: util/OptionsDB.cpp
Comment (adrianbroher) [inline util/OptionsDB.cpp]: Don't change code that ist not related to the topic.  Especially if it is just a style change.

Comment (memem359) [reply util/OptionsDB.cpp]: Except style change is entirely what this last commit was about (what Geoff just asked for).

This line is the main reason why you (and Geoff) had to keep pointing out style inconsistencies, because the original file (the example I had to go by) wasn't consistent.

If you would prefer a new PR, entirely for moving the bracket on this line, I will do that.
Or you can make an exception in this one particular case, since I was making a commit for style corrections in this file anyways.
(Your call.)

File: util/OptionsDB.cpp
Comment (memem359) [reply util/OptionsDB.cpp]: After you left your initial comment, I figured out what you just said.
I think it would help newer contributors to add those three sentences to the Programming wiki page.

This particular file didn't have other multi-line definitions, so I didn't know about that style convention until you mentioned it.
(The one exception being the declaration with a long initializer list, but my first guess was new line for { after initializer lists, not because it was multi-line.)
I've seen other (not FO) programmers who aways put { on the same line, even for multi-line definitions, so that choice may not be obvious.

File: util/OptionsDB.cpp
Comment (adrianbroher) [reply util/OptionsDB.cpp]: > Except style change is entirely what this last commit was about (what Geoff just asked for).
>
> This line is the main reason why you (and Geoff) had to keep pointing out style inconsistencies, because the original file (the example I had to go by) wasn't consistent.

The request was related to a specific changed line, not asking for any reformatting of the whole file.

> If you would prefer a new PR, entirely for moving the bracket on this line, I will do that.
> Or you can make an exception in this one particular case, since I was making a commit for style corrections in this file anyways.

I'm not interested in any of those, please undo that change.

Comment (memem359) [reply util/OptionsDB.cpp]: I'll do as you say. (Edit: Done. That particular line has been rolled back.)

Can you understand what I am pointing out?
This line does not conform to what you and Geoff want for the style (bracket on the same line as the definition). Leaving this unchanged means the next rookie programmer who comes across this file might also come to the (wrong) conclusion that the particular bracket stylistic choice is not important.

File: util/OptionsDB.cpp
Comment (geoffthemedio) [reply util/OptionsDB.cpp]: I prefer to put the `{` on a new line after a multi-line if expression or function header in order to make it clear at a glance where the actual code starts. Often it's unclear due to both being indented, and if the `{` is hidden off to the right somewhere, it isn't visible enough.

If it's a single-line function, then the `{` on the same line is fine, as I find it clearer what is code and what is definition or if test.

File: util/OptionsDB.cpp
Comment (geoffthemedio) [reply util/OptionsDB.cpp]: It should be fixed in a separate grooming commit. I'd be fine with it in the pull request, but so far as I understand his position, @adrianbroher perfers to keep grooming out of pull requests to make it easier to review what they actually substantially changed.

**Comment by geoffthemedio** [comment]
builds, seems to run OK, though I didn't test alternative option-setting mechanisms

File: util/OptionsDB.cpp
Comment (memem359) [reply util/OptionsDB.cpp]: I see the Wiki has changed.
I think that will help for future (novice) contributors.

In hindsight, I can appreciate the consistency that @geoffthemedio and @adrianbroher are setting for the code changes. At the time, the Wiki stated that I could do what I preferred, if it didn't conflict with the style of the existing code (which had opening brackets done in more than one way). I ended up creating a lot more work for myself, most of which would have been avoided if I had started with the new set of instructions instead of the old. Sorry about getting testy about my confusion.

Comment (adrianbroher) [reply util/OptionsDB.cpp]: > At the time, the Wiki stated that I could do what I preferred, if it didn't conflict with the style of the existing code (which had opening brackets done in more than one way).

There are several problems with this in general.

@geoffthemedio is right with his conclusion, topics and *grooming/housekeeping/whatever you want to call it* should be IMO separated at least by commits or even better left out completely from PRs. Why do I ask for this:

*  When looking back into the source code history via git to understand the implementation idea or asking the responsible developer why some code was done this way I don't want to step to some large number of unrelated grooming commits.
* A commit should be atomic to allow the most easy revert when needed. This is easier when a commit only holds a logical connected part. When mixing in changes that are unrelated (grooming or a different topic) this is way harder to to because the possibility for revert conflicts is higher.

Why are there are inconsistent styles: Well, it's old code and there was and is no technical enforcement of style that would prevent committing to the repository.  Also there is no consensus what well-formed code style should look like and how existing inconsistencies to that theoretical existing well-formed code style should be purged.  So the minimal agreement currently is:

* Use the style used in the same file.
* Don't add pointless grooming commits.
* Prefer fixing inconsistencies when touching a code line for other reasons.

**Review by adrianbroher — APPROVED** [review]

**Comment by adrianbroher** [comment]
@geoffthemedio dismissing your review, as you already wrote that this PR looks good to you.

**Comment by LGM-Doyle** [comment]
@memem359 merged 👍 .

**Pull request merged**
